nomad/eval_broker.go

```diff
@@ -559,6 +564,7 @@ func (b *EvalBroker) Ack(evalID, token string) error {
 		return fmt.Errorf("Token does not match for Evaluation ID")
 	}
 	jobID := unack.Eval.JobID
+	recentlyAckedIndex := unack.Eval.ModifyIndex
```

Comment:
**Update:** See my comment below about *canceling all but 1 eval.* I don't think the rest of this comment matters at all, but I spent like 4 hours on it so by golly I'm leaving it. :face_with_spiral_eyes: 

**tl;dr** -> I'm not sure this will cancel any evals. I think having the worker put `Eval.SnapshotIndex` into `structs.AckRequest` would allow the broker to cancel the maximum number of evals safely. This gets extremely difficult to reason about as it seems every layer of scheduler (broker, rpc, worker, scheduler) seems to do its own funny little index handling, so please check my work! :grimacing: 

AFAICT `Eval.ModifyIndex` does *not* include the index at which the eval was updated after committing a plan. `unack` is pulled out of `EvalBroker.unack` which contains the original copy of the eval from when it was first inserted into the broker as ready.

This means that `Eval.ModifyIndex` is pessimistic in the sense that it is the *minimum* possible value the Worker's snapshot was at when this evaluation was processed. Luckily we have a whole smorgasbord of indexes in the worker to choose from:

- `EvalDequeueResponse.WaitIndex` - the maximum Eval.ModifyIndex of all evals for this job when it was dequeued.
- `Eval.SnapshotIndex` - the _actual_ Raft index the worker's state snapshot used. *Only available on the _updated_ Eval.*
- `PlanResult.AllocIndex / PlanResponse.Index` - returned by the `Plan.Submit` RPC to the worker who submitted the plan, but never used. I think the worker only ever uses the `PlanResult.RefreshIndex` and _that_ index never bubbles up to where the Worker calls `Ack`.

I think the following holds as an invariant:

```
PlanResult.AllocIndex > {Updated}Eval.SnapshotIndex >= EvalDequeueResponse.WaitIndex >= {Broker}Eval.ModifyIndex
          ^                          ^                               ^                              ^
          |                          |                               |                              |
          |    no point in processing new evals < this index         |                 this is what's being used
          |                                                          |
too new! cluster may have changed prior to this                      |
                                                                     |      
     safe choice as if it's > ModifyIndex that indicates there are pending evals that can be optimized out!

```

I think `Eval.SnapshotIndex` is precisely what we're looking for. There's no point processing a new eval for the same job if we're not going to process it at a higher index than this because we will come up with the same result as before since we haven't allowed the cluster state to change!

Since there could be updates between `PR.AllocIndex` and `SnapshotIndex`, we can't drop any evals based on `AllocIndex`.